core/graphelier-service/api/hndlrs/obhandler.go

```diff
@@ -6,47 +6,119 @@ import (
 	"strconv"
 
 	"graphelier/core/graphelier-service/models"
+	"graphelier/core/graphelier-service/utils"
 
 	"github.com/gorilla/mux"
 )
 
-// JSONOrderbook : Sends an orderbook as json
-func JSONOrderbook(env *Env, w http.ResponseWriter, r *http.Request) error {
+// FetchOrderbook : Sends an orderbook state based on instrument and timestamp
+func FetchOrderbook(env *Env, w http.ResponseWriter, r *http.Request) error {
 	params := mux.Vars(r)
 
 	instrument := params["instrument"]
-	timestamp := params["timestamp"]
+	t := params["timestamp"]
 
 	// Convert string to uint64
-	intTimestamp, err := strconv.ParseUint(timestamp, 10, 64)
+	timestamp, err := strconv.ParseUint(t, 10, 64)
 	if err != nil {
-		return err
+		return StatusError{400, err}
 	}
-	var orderbook *models.Orderbook
-	var messages []*models.Message
 
-	if divisor := uint64(10000000000); intTimestamp%divisor != 0 {
-		latestFullSnapshot := intTimestamp / divisor * divisor
-		orderbook, err = env.Connector.GetOrderbook(instrument, intTimestamp)
-		if err != nil {
-			return err
+	orderbook, err := env.Connector.GetOrderbook(instrument, timestamp)
+	if err != nil {
+		return StatusError{500, err}
+	}
+	messages, err := env.Connector.GetMessagesByTimestamp(instrument, timestamp)
+	if err != nil {
+		return StatusError{500, err}
+	}
+	orderbook.ApplyMessagesToOrderbook(messages)
+
+	w.WriteHeader(http.StatusOK)
+	err = json.NewEncoder(w).Encode(orderbook)
+	if err != nil {
+		return StatusError{500, err}
+	}
+
+	w.WriteHeader(http.StatusOK)
+
+	return nil
+}
+
+// FetchOrderbookDelta : Sends an orderbook state based on the instrument, sod offset, and message offset
+func FetchOrderbookDelta(env *Env, w http.ResponseWriter, r *http.Request) (err error) {
+	params := mux.Vars(r)
+
+	instrument := params["instrument"]
+	so := params["sod_offset"]
+	nm := params["num_messages"]
+
+	// Convert strings to int64
+	sodOffset, err := strconv.ParseInt(so, 10, 64)
+	if err != nil {
+		return StatusError{400, err}
+	}
+	numMessages, err := strconv.ParseInt(nm, 10, 64)
+	if err != nil {
+		return StatusError{400, err}
+	}
+
+	sodMessage, err := env.Connector.GetSingleMessage(instrument, sodOffset)
+	if err != nil {
+		return StatusError{500, err}
+	}
+	orderbook, err := env.Connector.GetOrderbook(instrument, sodMessage.Timestamp)
+	if err != nil {
+		return StatusError{500, err}
+	}
+	totalOffset := int64(int64(sodMessage.SodOffset-orderbook.LastSodOffset) + numMessages)
+
+	// get previous book for high number of backward messages
+	for totalOffset < 0 {
+		prevSnapTime := int64(orderbook.Timestamp - 10000000000)
```

Comment:
If uint becomes negative you'll overflow to max int instead of negative numbers
```suggestion
		prevSnapTime := int64(orderbook.Timestamp) - 10000000000
```

Reply:
there's no uint here?

Reply:
this is being set to an int so there's no overflow here and the number will actually be negative. this cast makes all numbers inside it ints since uint - int is not possible